### PR TITLE
Pricefeed v2

### PR DESCRIPTION
**Pull request by spider-g**

### CODE REVIEW

**Review by grok13 — CHANGES_REQUESTED** [review]
Some small comments, but great work!

File: contracts/PriceFeed.sol [hunk 1]
```diff
@@ -41,232 +46,149 @@ contract PriceFeed is IPriceFeed, OwnableUpgradeable, UUPSUpgradeable, BaseMath,
...
-			uint256 timestamp,
+			uint256 updatedAt,
 			uint80 /* answeredInRound */
 		) {
-			// If call to Chainlink succeeds, return the response and success = true
-			response.roundId = roundId;
-			response.answer = answer;
-			response.timestamp = timestamp;
-			response.success = true;
+			if (roundId != 0 && updatedAt != 0 && answer != 0 && updatedAt <= block.timestamp) {
```
[the remark this answers is not in the document]
Comment (spider-g) [reply contracts/PriceFeed.sol:149]: D'accord.

File: contracts/PriceFeed.sol [hunk 2]
```diff
@@ -41,232 +46,149 @@ contract PriceFeed is IPriceFeed, OwnableUpgradeable, UUPSUpgradeable, BaseMath,
...
-			_checkTimelock(); // Subsequent updates need to go through the timelock contract
-		}
-
-		if (
-			_maxDeviationBetweenRounds < MAX_PRICE_DEVIATION_BETWEEN_ROUNDS_LOWER_LIMIT ||
-			_maxDeviationBetweenRounds > MAX_PRICE_DEVIATION_BETWEEN_ROUNDS_UPPER_LIMIT
-		) {
-			revert PriceFeed__InvalidPriceDeviationParamError();
+		_requireOwnerOrTimelock(_token, _isFallback);
+		if (_isFallback && oracles[_token].oracleAddress == address(0)) {
```
[the remark this answers is not in the document]
Comment (spider-g) [reply contracts/PriceFeed.sol:56]: I refrained from adding too many input checks as this is an admin-only function. I'd like to believe we know what we're doing :)

File: contracts/PriceFeed.sol [hunk 3]
```diff
@@ -41,232 +46,149 @@ contract PriceFeed is IPriceFeed, OwnableUpgradeable, UUPSUpgradeable, BaseMath,
...
-			if (oracle.isFeedWorking) {
-				_updateFeedStatus(_token, oracle, false);
-			}
-			PriceRecord memory priceRecord = priceRecords[_token];
-			if (_isPriceStale(priceRecord.timestamp)) {
-				revert PriceFeed__FeedFrozenError(_token);
-			}
-			return priceRecord.scaledPrice;
+	function _fetchDecimals(address _oracle, ProviderType _type) internal view returns (uint8) {
+		if (_type == ProviderType.Chainlink) {
```
[the remark this answers is not in the document]
Comment (spider-g) [reply contracts/PriceFeed.sol:114]: ATM we only support the Chainlink interface, but the enum type was left for the future, as upgrading a struct can be tricky. Depending on the upcoming additions, we can code a different routine or add an adapter contract that implements the CL Aggregator interface.

**Pull request merged**